api/test/wfl/unit/source_test.clj

```diff
@@ -33,3 +33,54 @@
           actual       (#'source/result-or-catch callable)]
       (is (= status (:status actual)))
       (is (= body-message (get-in actual [:body :message]))))))
+
+(deftest test-filter-and-combine-tdr-source-details
+  (let [time           "2022-01-21T08:%s:00.000000Z"
+        running-rows   ["r1" "r2"]
+        running        {:snapshot_creation_job_status "running"
+                        :datarepo_row_ids             running-rows
+                        :start_time                   (format time "00")
+                        :end_time                     (format time "10")}
+        succeeded-rows ["s1" "s2"]
+        succeeded      {:snapshot_creation_job_status "succeeded"
+                        :datarepo_row_ids             succeeded-rows
+                        :start_time                   (format time "20")
+                        :end_time                     (format time "30")}
+        failed-rows    ["f1" "f2"]
+        failed         {:snapshot_creation_job_status "failed"
+                        :datarepo_row_ids             failed-rows
+                        :start_time                   (format time "40")
+                        :end_time                     (format time "50")}]
+    (testing "No rows"
+      (is (nil? (#'source/filter-and-combine-tdr-source-details
+                 []))
+          "Empty rows passed in should not break processing"))
+    (testing "Keep all rows"
+      (let [{:keys [datarepo_row_ids start_time end_time]}
+            (#'source/filter-and-combine-tdr-source-details
+             [succeeded])]
+        (is (= succeeded-rows datarepo_row_ids)
+            "succeeded rows should be kept")
+        (is (= (:start_time succeeded) start_time))
+        (is (= (:end_time succeeded) end_time))))
+    (testing "Keep some rows"
+      (let [{:keys [datarepo_row_ids start_time end_time]}
+            (#'source/filter-and-combine-tdr-source-details
+             [failed succeeded])]
+        (is (= succeeded-rows datarepo_row_ids)
+            "succeeded rows should be kept")
+        (is (= (:start_time succeeded) start_time))
+        (is (= (:end_time succeeded) end_time)))
+      (let [{:keys [datarepo_row_ids start_time end_time]}
+            (#'source/filter-and-combine-tdr-source-details
+             [running succeeded failed])]
+        (is (= (concat running-rows succeeded-rows) datarepo_row_ids)
+            "running and succeeded rows should be kept")
+        (is (= (:start_time running) start_time)
+            "Earliest start time should be taken")
+        (is (= (:end_time succeeded) end_time)
+            "Latest end time should be taken")))
+    (testing "Keep no rows"
+      (is (nil? (#'source/filter-and-combine-tdr-source-details
```

Comment:
This is the case where we experienced the bug, more info in the ticket description.

Reply:
Because `failed` is too old or because it has only `failed-rows`?
(The latter, right?)
Then is there a "snapshot request too old but has succeeded rows" case?

Reply:
Correct, the latter.  If rows are too old, either way they will not get considered for reprocessing (expected behavior.)